tests/DecisionServiceTests/DecisionServiceTest.php

```diff
@@ -603,6 +614,445 @@ public function testGetVariationWithBucketingId()
+            'variation' => $variation
+        ];
+        $feature_flag = $this->config->getFeatureFlagFromKey('boolean_feature');
+        $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::INFO, 
+            "The user 'user_1' is bucketed into experiment 'group_experiment_1' of feature 'boolean_feature'.");
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeatureExperiment($feature_flag, 'user_1', []),
+            $expected_decision
+        );
+
+    }
+
+     // should return nil and log a message when the user is not bucketed into any of the mutex experiments      
+    public function testGetVariationForFeatureExperimentGivenMutexGroupAndUserNotBucketed(){
+        $mutex_exp = $this->config->getExperimentFromKey('group_experiment_1');
+        $variation = $mutex_exp->getVariations()[0];
+        $this->decisionServiceMock->expects($this->at(0))
+        ->method('getVariation')
+        ->will($this->returnValue(null));
+
+
+        $mutex_exp = $this->config->getExperimentFromKey('group_experiment_1');
+        $feature_flag = $this->config->getFeatureFlagFromKey('boolean_feature');
+        $this->loggerMock->expects($this->at(0))
+        ->method('log')
+        ->with(Logger::INFO, 
+            "The user 'user_1' is not bucketed into any of the experiments on the feature 'boolean_feature'.");
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeatureExperiment($feature_flag, 'user_1', []),
+            null
+        );
+    }
+
+    //should return the bucketed experiment and variation
+    public function testGetVariationForFeatureWhenTheUserIsBucketedIntoFeatureExperiment(){
+
+       $this->decisionServiceMock = $this->getMockBuilder(DecisionService::class)
+       ->setConstructorArgs(array($this->loggerMock, $this->config))
+       ->setMethods(array('getVariationForFeatureExperiment'))
+       ->getMock();
+
+       $feature_flag = $this->config->getFeatureFlagFromKey('string_single_variable_feature');
+       $expected_experiment_id = $feature_flag->getExperimentIds()[0];
+       $expected_experiment = $this->config->getExperimentFromId($expected_experiment_id);
+       $expected_variation = $expected_experiment->getVariations()[0];
+       $expected_decision = [
+        'experiment' => $expected_experiment,
+        'variation' => $expected_variation
+        ];
+
+        $this->decisionServiceMock->expects($this->at(0))
+        ->method('getVariationForFeatureExperiment')
+        ->will($this->returnValue($expected_decision));
+
+        $this->assertEquals(
+            $this->decisionServiceMock->getVariationForFeature($feature_flag, 'user_1', []),
+            $expected_decision
+        );
+    }
+
+    // should return the bucketed variation and null experiment
+    public function testGetVariationForFeatureWhenTheUserIsNotBucketedIntoFeatureExperimentAndBucketedToFeatureRollout(){
```

Comment:
can we make shorter method name?
testGetVariationForFeatureWhenBucketedToFeatureRollout?

Reply:
it is obvious that if the feature is not bucketed iinto feature experiment, it must go into the feature rollout.

Reply:
Done